File: react-native-pytorch-core/android/src/main/java/org/pytorch/rn/core/ml/processing/unpacker/TensorToImageUnpacker.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package org.pytorch.rn.core.ml.processing.unpacker;

import android.graphics.Bitmap;
import android.graphics.Color;
import com.facebook.react.bridge.WritableMap;
import org.json.JSONException;
import org.json.JSONObject;
import org.pytorch.IValue;
import org.pytorch.Tensor;
import org.pytorch.rn.core.image.IImage;
import org.pytorch.rn.core.image.Image;
import org.pytorch.rn.core.javascript.JSContext;
import org.pytorch.rn.core.ml.processing.PackerContext;
import org.pytorch.rn.core.ml.processing.Unpacker;

public class TensorToImageUnpacker implements Unpacker {

  private static final String KEY_KEY = "key";

  @Override
  public void unpack(
      IValue ivalue, JSONObject jobject, WritableMap result, PackerContext packerContext)
      throws JSONException {
    Tensor tensor = ivalue.toTensor();
    Bitmap bitmap = tensorToBitmap(tensor);
    IImage image = new Image(bitmap);
    JSContext.NativeJSRef ref = JSContext.wrapObject(image);

    final String key = jobject.getString(KEY_KEY);
    result.putMap(key, ref.getJSRef());
  }

  private Bitmap tensorToBitmap(Tensor tensor) {
    float[] data = tensor.getDataAsFloatArray();
    long[] shape = tensor.shape();
    int width = (int) shape[3];
    int height = (int) shape[2];

    Bitmap bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);

    // Determine the min/max value of data
    float max = 0;
    float min = Float.MAX_VALUE;
    for (float f : data) {
      if (f > max) {
        max = f;
      }
      if (f < min) {
        min = f;
      }
    }

    int delta = (int) (max - min);
    for (int i = 0; i < width * height; i++) {
      int r = (int) ((data[i] - min) / delta * 255.0f);
      int g = (int) ((data[i + width * height] - min) / delta * 255.0f);
      int b = (int) ((data[i + width * height * 2] - min) / delta * 255.0f);

      int x = i / height;
      int y = i % width;

      int color = Color.rgb(r, g, b);
      bitmap.setPixel(x, y, color);
    }
    return bitmap;
  }
}

File: react-native-pytorch-core/ios/ML/Processing/Unpacker/TensorToImageUnpacker.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation
import SwiftyJSON

class TensorToImageUnpacker: Unpacker {
    func unpack(ivalue: IValue, modelSpec: JSON, result: inout [String: Any], packerContext: PackerContext) throws {
        let unpack = modelSpec["unpack"]

        guard let key = unpack["key"].string else {
            throw BaseIValuePackerError.missingKeyParam
        }

        guard let tensor = ivalue.toTensor() else {
            throw BaseIValuePackerError.invalidUnpackType
        }
        guard let cgImage = try tensorToBitmap(tensor: tensor) else {
            throw BaseIValuePackerError.invalidUnpackType
        }
        let image = Image(image: cgImage)
        let ref = JSContext.wrapObject(object: image).getJSRef()
        result[key] = ref
    }

    private func tensorToBitmap(tensor: Tensor) throws -> CGImage? {
        guard let data = tensor.getDataAsArray() as? [Float32] else {
            throw BaseIValuePackerError.encodeImageError
        }
        let shape = tensor.shape

        let width = shape[3].intValue
        let height = shape[2].intValue

        // Determine the min/max value of data
        var max: Float32 = 0
        var min: Float32 = Float32.greatestFiniteMagnitude
        for val in data {
            if val > max {
              max = val
            }
            if val < min {
              min = val
            }
        }

        let delta = (max - min)
        var rgba: [UInt8] = [UInt8](repeating: 0, count: 4 * width * height)
        for val in 0..<(width * height) {
            rgba[4 * val] = (UInt8) ((data[val] - min) / delta * 255)
            rgba[4 * val + 1] = (UInt8) ((data[width * height + val] - min) / delta * 255)
            rgba[4 * val + 2] = (UInt8) ((data[width * height * 2 + val] - min) / delta * 255)
            rgba[4 * val + 3] = 255
        }

        let dataForProvider: Data = Data(rgba)
        let bytesPerPixel = 4
        let bitsPerPixel = 32
        let bytesPerRow = bytesPerPixel * width
        let bitsPerComponent = 8
        let bitmapInfo = CGBitmapInfo(rawValue: CGImageAlphaInfo.premultipliedLast.rawValue)
        let provider: CGDataProvider! = CGDataProvider(data: dataForProvider as CFData)

        let cgImage = CGImage(
            width: width,
            height: height,
            bitsPerComponent: bitsPerComponent,
            bitsPerPixel: bitsPerPixel,
            bytesPerRow: bytesPerRow,
            space: CGColorSpaceCreateDeviceRGB(),
            bitmapInfo: bitmapInfo,
            provider: provider,
            decode: nil,
            shouldInterpolate: false,
            intent: .defaultIntent
        )
        return cgImage
    }
}
